File: qbopt/frontend/raising_float_values.py
"""Resolve self-contained floating stacks to ordinary MIR value operands."""

from dataclasses import replace

from qbopt.frontend import fpstack
from qbopt.model import mir
from qbopt.analysis import ssa


def _regions(block):
    """Balanced typed sequences, separated by calls or unknown stack effects."""
    run = []
    depth = 0
    for op in block.ops:
        if op.barrier or op.kind is mir.Kind.CALL or (op.stack is not None and op.floating is None):
            run, depth = [], 0
            continue
        run.append(op)
        if op.stack is not None:
            depth += op.stack
            if depth < 0:
                run, depth = [], 0
                continue
        if depth == 0:
            if any(one.stack is not None for one in run):
                yield tuple(run)
            run = []


_ARITHMETIC = (mir.Kind.FADD, mir.Kind.FSUB, mir.Kind.FMUL, mir.Kind.FDIV)


def loaded(body: mir.MirBody) -> mir.MirBody:
    """Arithmetic over values: a memory operand becomes its own load, which carries the format."""
    from qbopt.model import ir
    from qbopt.model.floating import Format
    from qbopt.model.floating import Precision
    from qbopt.model.floating import Rounding
    from qbopt.model.floating import Semantics

    values = tuple(ssa.values(body))
    serial = max((value.id for value in values), default=0)
    variable = max((value.variable for value in values), default=0)
    blocks = []
    for block in body.blocks:
        ops = []
        for op in block.ops:
            if (
                op.kind not in _ARITHMETIC
                or op.floating is None
                or op.stores
                or len(op.loads) != 1
                or len(op.args) != 2
                or not isinstance(op.args[0], mir.Held)
                or op.args[0].width != 10
                or not isinstance(op.args[1], mir.Cell)
            ):
                ops.append(op)
                continue
            kept, cell = op.args
            serial += 1
            variable += 1
            read = mir.Held(mir.Value(serial, op.at, variable=variable, version=1), 10)
            integer = op.name.startswith("fi")
            ops.append(
                mir.detached(
                    op,
                    kind=mir.Kind.FLOAD,
                    op=ir.Operation.FLOAT_LOAD,
                    name="fild" if integer else "fld",
                    args=(cell,),
                    results=(read,),
                    defines=(read.value,),
                    uses=tuple(value for value in op.uses if value != kept.value),
                    merges={},
                    floating=Semantics(
                        (op.floating.inputs[1],),
                        Format.EXTENDED80,
                        Precision.EXACT,
                        Rounding.NONE,
                        op.floating.exceptions,
                    ),
                    floating_origin=None,
                    stack=None,
                    raised=None,
                    symbol=None,
                    covers=(op.at, op.at),
                    extra_covers=(),
                    id=next(mir._IDS),
                )
            )
            ops.append(
                replace(
                    op,
                    name="f" + op.name[2:] if integer else op.name,
                    args=(kept, read),
                    uses=(kept.value, read.value),
                    loads=(),
                    raised=None,
                    floating=replace(op.floating, inputs=(Format.EXTENDED80, Format.EXTENDED80)),
                )
            )
        blocks.append(replace(block, ops=tuple(ops)))
    return replace(body, blocks=tuple(blocks))


def raised(body: mir.MirBody) -> mir.MirBody:
    values = tuple(ssa.values(body))
    serial = max((value.id for value in values), default=0)
    variable = max((value.variable for value in values), default=0)
    blocks = []
    for block in body.blocks:
        regions = tuple(_regions(block))
        operations = [op for region in regions for op in region if op.stack is not None]
        readings = {}
        for region in regions:
            readings.update(fpstack.readings(replace(body, blocks=(replace(block, ops=region),))))
        if (not operations or any(op.floating is None for op in operations)
            or len({op.at for op in operations}) != len(operations)
            or sum(op.stack for op in operations) != 0):
            blocks.append(block)
            continue
        known = set()
        valid = True
        for op in operations:
            reading = readings.get(op.at)
            if reading is None or any(value not in known for value in reading.uses.values()):
                valid = False
                break
            if any(isinstance(arg, mir.Opaque) for arg in op.results) and reading.defines is None:
                valid = False
                break
            if reading.defines is not None:
                known.add(reading.defines)
        if not valid:
            blocks.append(block)
            continue
        held = {}
        for value in sorted(known, key=lambda one: one.id):
            serial += 1
            variable += 1
            held[value] = mir.Held(mir.Value(serial, value.at, variable=variable, version=1), 10)
        sequence = tuple(op.at for op in operations)
        selected = {id(op) for op in operations}
        ops = []
        for op in block.ops:
            if id(op) not in selected:
                ops.append(op)
                continue
            reading = readings[op.at]
            def source(arg):
                if isinstance(arg, mir.Opaque) and arg.name.startswith("st") and arg.name[2:].isdigit():
                    return held[reading.uses[int(arg.name[2:])]]
                return arg
            args = tuple(source(arg) for arg in op.args)
            results = tuple(held[reading.defines] if isinstance(arg, mir.Opaque) else arg for arg in op.results)
            baseline = mir.FloatingOrigin(block.at, sequence, op.at, op.kind, op.floating,
                                          args, results, op.args, op.results)
            uses = tuple(dict.fromkeys((*op.uses, *(arg.value for arg in args if isinstance(arg, mir.Held)))))
            defines = tuple(dict.fromkeys((*op.defines, *(arg.value for arg in results if isinstance(arg, mir.Held)))))
            ops.append(replace(op, args=args, results=results, uses=uses, defines=defines, floating_origin=baseline))
        blocks.append(replace(block, ops=tuple(ops)))
    return replace(body, blocks=tuple(blocks))

File: crates/llrm-core/src/frontends/bc/raising_float_values.rs
//! Port of `qbopt/frontend/raising_float_values.py`: resolve self-contained
//! floating stacks to ordinary MIR value operands.

use std::collections::BTreeSet;

use crate::analysis::ssa;
use crate::frontends::bc::fpstack::{self, Float, Reading};
use crate::model::floating::{Format, Precision, Rounding, Semantics};
use crate::model::ir::Operation;
use crate::model::mir::{self, Arg, FloatingOrigin, Held, Kind, MirBlock, Op, OpCode, OrderedMap, RaisedBody, Value};
use crate::support::hash::{IndexMap, IndexSet};
use crate::support::pyset::PySet;

/// Balanced typed sequences, separated by calls or unknown stack effects,
/// as indices into `block.ops`.
fn _regions(block: &MirBlock) -> Vec<Vec<usize>> {
    let mut out = Vec::new();
    let mut run: Vec<usize> = Vec::new();
    let mut depth = 0;
    for (index, op) in block.ops.iter().enumerate() {
        if op.barrier()
            || op.kind == Kind::Call
            || (op.stack.is_some() && op.floating.is_none() && op.kind != Kind::Fcompare)
        {
            (run, depth) = (Vec::new(), 0);
            continue;
        }
        run.push(index);
        if let Some(stack) = op.stack {
            depth += stack;
            if depth < 0 {
                (run, depth) = (Vec::new(), 0);
                continue;
            }
        }
        if depth == 0 {
            if run.iter().any(|&one| block.ops[one].stack.is_some()) {
                out.push(run.clone());
            }
            run = Vec::new();
        }
    }
    out
}

const _ARITHMETIC: [Kind; 4] = [Kind::Fadd, Kind::Fsub, Kind::Fmul, Kind::Fdiv];

/// Arithmetic over values: a memory operand becomes its own load, which
/// carries the format.
pub fn loaded(body: RaisedBody) -> RaisedBody {
    let values: Vec<Value> = ssa::values(&body).collect();
    let mut serial = values.iter().map(|value| value.id).max().unwrap_or(0);
    let mut variable = values.iter().map(|value| value.variable).max().unwrap_or(0);
    let mut blocks = Vec::new();
    for block in &body.blocks {
        let mut ops = Vec::new();
        for op in &block.ops {
            let (Some(floating), [Arg::Held(kept), Arg::Cell(cell)]) = (&op.floating, op.args.as_slice()) else {
                ops.push(op.clone());
                continue;
            };
            if !_ARITHMETIC.contains(&op.kind) || !op.stores.is_empty() || op.loads.len() != 1 || kept.width != 10 {
                ops.push(op.clone());
                continue;
            }
            serial += 1;
            variable += 1;
            let read = Held { value: Value { variable, version: 1, ..Value::new(serial, op.at) }, width: 10 };
            let integer = op.name.starts_with("fi");
            let mut load = op.clone();
            load.kind = Kind::Fload;
            load.op = Some(OpCode::Operation(Operation::FloatLoad));
            load.name = if integer { "fild" } else { "fld" }.to_owned();
            load.args = vec![Arg::Cell(cell.clone())];
            load.results = vec![Arg::Held(read)];
            load.defines = vec![read.value];
            load.uses = op.uses.iter().copied().filter(|value| *value != kept.value).collect();
            load.merges = OrderedMap::new();
            load.floating = Some(Semantics::with_exceptions(
                [floating.inputs[1]],
                Format::Extended80,
                Precision::Exact,
                Rounding::None,
                floating.exceptions,
            ));
            load.floating_origin = None;
            load.stack = None;
            load.raised = None;
            load.symbol = None;
            load.source = Some(mir::next_id());
            ops.push(mir::source_free(load));
            let mut arithmetic = op.clone();
            if integer {
                arithmetic.name = format!("f{}", &op.name[2..]);
            }
            arithmetic.args = vec![Arg::Held(*kept), Arg::Held(read)];
            arithmetic.uses = vec![kept.value, read.value];
            arithmetic.loads = Vec::new();
            arithmetic.raised = None;
            arithmetic.floating = Some(Semantics {
                inputs: [Format::Extended80, Format::Extended80].into(),
                ..floating.clone()
            });
            ops.push(arithmetic);
        }
        blocks.push(block.with_ops(ops));
    }
    body.with_blocks(blocks)
}

pub fn raised(body: RaisedBody) -> RaisedBody {
    let values: Vec<Value> = ssa::values(&body).collect();
    let mut serial = values.iter().map(|value| value.id).max().unwrap_or(0);
    let mut variable = values.iter().map(|value| value.variable).max().unwrap_or(0);
    let mut blocks = Vec::new();
    for block in &body.blocks {
        let regions = _regions(block);
        let operations: Vec<usize> =
            regions.iter().flatten().copied().filter(|&index| block.ops[index].stack.is_some()).collect();
        let mut readings: IndexMap<i64, Reading> = IndexMap::default();
        for region in &regions {
            let alone = block.with_ops(region.iter().map(|&index| block.ops[index].clone()).collect());
            readings.extend(fpstack::readings(&body.body.with_blocks(vec![alone])));
        }
        let ops_of = || operations.iter().map(|&index| &block.ops[index]);
        if operations.is_empty()
            || ops_of().any(|op| op.floating.is_none() && op.kind != Kind::Fcompare)
            || ops_of().map(|op| op.at).collect::<BTreeSet<_>>().len() != operations.len()
            || ops_of().map(|op| op.stack.unwrap()).sum::<i64>() != 0
        {
            blocks.push(block.clone());
            continue;
        }
        let mut known: PySet<Float> = PySet::new();
        let mut valid = true;
        for op in ops_of() {
            let reading = readings.get(&op.at);
            let Some(reading) = reading.filter(|reading| reading.uses.values().all(|value| known.contains(value)))
            else {
                valid = false;
                break;
            };
            if op.results.iter().any(|arg| matches!(arg, Arg::Opaque(_))) && reading.defines.is_none() {
                valid = false;
                break;
            }
            if let Some(defines) = reading.defines {
                known.add(defines);
            }
        }
        if !valid {
            blocks.push(block.clone());
            continue;
        }
        let mut held: IndexMap<Float, Held> = IndexMap::default();
        let mut sorted: Vec<Float> = known.iter().copied().collect();
        sorted.sort_by_key(|one| one.id);
        for value in sorted {
            serial += 1;
            variable += 1;
            let made = Value { variable, version: 1, ..Value::new(serial, value.at.unwrap()) };
            held.insert(value, Held { value: made, width: 10 });
        }
        let sequence: Vec<i64> = ops_of().map(|op| op.at).collect();
        let selected: BTreeSet<usize> = operations.iter().copied().collect();
        let mut ops = Vec::new();
        for (index, op) in block.ops.iter().enumerate() {
            if !selected.contains(&index) {
                ops.push(op.clone());
                continue;
            }
            let reading = &readings[&op.at];

            let source = |arg: &Arg| -> Arg {
                if let Arg::Opaque(one) = arg {
                    if let Some(digits) = one.name.strip_prefix("st") {
                        if !digits.is_empty() && digits.chars().all(|c| c.is_ascii_digit()) {
                            return Arg::Held(held[&reading.uses[&digits.parse::<i64>().unwrap()]]);
                        }
                    }
                }
                arg.clone()
            };

            let args: Vec<Arg> = op.args.iter().map(source).collect();
            let results: Vec<Arg> = op
                .results
                .iter()
                .map(|arg| match arg {
                    Arg::Opaque(_) => Arg::Held(held[&reading.defines.unwrap()]),
                    _ => arg.clone(),
                })
                .collect();
            let baseline = if op.kind == Kind::Fcompare {
                None
            } else {
                Some(FloatingOrigin {
                    block: block.at,
                    sequence: sequence.clone(),
                    at: op.at,
                    kind: op.kind,
                    semantics: op.floating.clone().unwrap(),
                    inputs: args.clone(),
                    outputs: results.clone(),
                    machine_inputs: op.args.clone(),
                    machine_outputs: op.results.clone(),
                })
            };
            let held_values = |operands: &[Arg]| -> Vec<Value> {
                operands
                    .iter()
                    .filter_map(|arg| match arg {
                        Arg::Held(one) => Some(one.value),
                        _ => None,
                    })
                    .collect()
            };
            let mut made: Op = op.clone();
            made.uses = op.uses.iter().copied().chain(held_values(&args)).collect::<IndexSet<_>>().into_iter().collect();
            made.defines =
                op.defines.iter().copied().chain(held_values(&results)).collect::<IndexSet<_>>().into_iter().collect();
            made.args = args;
            made.results = results;
            made.floating_origin = baseline;
            ops.push(made);
        }
        blocks.push(block.with_ops(ops));
    }
    body.with_blocks(blocks)
}
